SortByKey and GroupByKey: Invalid input.

aosyatnik opened the pull request:
### Purpose

Link to YouTrack:
[MAGN-8753](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8753) Add 'Invalid Input' error message to SortByKey and GroupByKey

I thought that would be the easiest task. I was wrong.

I created SortByKey and GroupByKey as UI nodes, they were called from FunctionObject.ds file. But to show any warning we need throw exception. In design script they is no way of creating exception.

So, I had to make SortByKey and GroupByKey as zero touch nodes. Also I duplicated old tests, but rewrote them as for zero touch node tests.

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@mjkkirschner 

### FYIs

@Racel

aosyatnik commented:
All tests passed.
![image](https://cloud.githubusercontent.com/assets/8158404/10481019/ace15ff6-7276-11e5-99b9-06665f06f3d0.png)

mjkkirschner commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -235,6 +236,87 @@ public static IDictionary Deconstruct(IList list)
+            // First items with number keys, then items with letter keys.
+            var sortedPairs = numberKeyPairs.Concat(keyPairs);
+
+            var sortedList = sortedPairs.Select(x => x.item);
```
I think it might be best to convert ToList() at this point before returning, this is done with the group by nodes and other list nodes return lists not ienums.

mjkkirschner commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -235,6 +236,87 @@ public static IDictionary Deconstruct(IList list)
+            }
+
+            if (list.Count != keys.Count)
+                return null;
```
it's been so many iterations I have forgotten :), is there a reason why we just return null here and don't have a warning message explaining why?

aosyatnik replied:
There wasn't any special reason why we didn't show warning. The only reason why we didn't do it, because it was design script and in design there is no opportunity to generate warning. At this place I can throw exception with some clear description. 
@Racel , would you like to show warning, if number of list items doesn't equal number of key items? If you do want, can you provide appropriate message?
Thanks.

mjkkirschner commented on test/Libraries/CoreNodesTests/ListTests.cs:
```diff
@@ -950,5 +950,206 @@ public static void Chop6()
             Assert.Throws<InvalidCastException>(() => { List.Chop(list, lengths); });
         }
+
+        [Test]
```
are there any tests for SortByFunction? - are these covered somewhere else in the tests? If not can we add some?

aosyatnik replied:
I added additional test for GrouByFunction. Test for SortByFunction was created before me...

Racel commented:
@aosyatnik - when # of items does not equal # of keys, please provide the warning: "Number of items does not match the number of keys"

Racel commented:
@aosyatnik - also, see my comment in Youtrack. Can we change the wording of the invalid input warning? I have added the new text to the description in Youtrack.

aosyatnik commented:
@mjkkirschner , I updated PR, could you take a look?

mjkkirschner commented on src/Libraries/CoreNodeModels/Properties/Resources.resx:
```diff
@@ -584,40 +584,4 @@
     <value>Points</value>
   </data>
-  <data name="ListGroupByKeyDescription" xml:space="preserve">
-    <value>Group items into sub-lists based on their like key values</value>
```
sorry I missed this the first time around - is there a reason you have not added all of these descriptions and tooltips to the zero touch nodes?

mjkkirschner commented on src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds:
```diff
@@ -479,33 +479,7 @@ def SortByFunction(list : var[]..[], func:_FunctionObject)
         {
             _keys[_index] = __Apply(func, list[_index]);
         }
-        return = Sorting.sortByKey(list, _keys)[0];
```
another concern I have is this file:
https://github.com/DynamoDS/Dynamo/blob/11f90d72e1047fcea9ece3818e90cf420890864d/src/DynamoCore/BuiltInAndOperators/BuiltIn.Migrations.xml

just a question for you: does this migration work correctly now that sortbykey and group by key are zero touch nodes?

mjkkirschner commented:
@aosyatnik this looks really good, thanks for your work!

Theres one more issue with the zero touch node losing the nice strings we had to tooltips and descriptions from the other nodes, and just a concern about the migration file that I would like you to check out. Other than that looks good to merge.

aosyatnik commented:
@mjkkirschner migration for BuiltIn nodes works properly. I added migration for UI nodes (SortByKey and GroupByKey).
We don't lose descriptions. I added them as xml comment <summary>.
![image](https://cloud.githubusercontent.com/assets/8158404/10538014/5331b120-73fe-11e5-88dd-9c2cb43a75d4.png) 
![image](https://cloud.githubusercontent.com/assets/8158404/10538023/60530278-73fe-11e5-85ca-4381bb0aec28.png)

aosyatnik commented:
Passed. 
![image](https://cloud.githubusercontent.com/assets/8158404/10539441/3f7f2dd8-7408-11e5-84a8-a71b0851d5e6.png)
